modules/auxiliary/scanner/ssh/ssh_login_pubkey.rb

```diff
@@ -207,26 +210,43 @@ def realm
     def valid?
       @error_list = []
       @key_data = Set.new
-      if File.directory?(@key_path)
-        @key_files ||= Dir.entries(@key_path).reject { |f| f =~ /^\x2e|\x2epub$/ }
-        @key_files.each do |f|
+
+      if @key_path.present?
+        if File.directory?(@key_path)
+          @key_files ||= Dir.entries(@key_path).reject { |f| f =~ /^\x2e|\x2epub$/ }
+          @key_files.each do |f|
+            begin
+              data = read_key(File.join(@key_path, f))
+              @key_data << data if valid_key?(data)
+            rescue StandardError => e
+              @error_list << "#{File.join(@key_path, f)}: #{e}"
+            end
+          end
+        elsif File.file?(@key_path)
           begin
-            data = read_key(File.join(@key_path, f))
+            data = read_key(@key_path)
             @key_data << data if valid_key?(data)
           rescue StandardError => e
-            @error_list << "#{File.join(@key_path, f)}: #{e}"
+            @error_list << "#{@key_path} could not be read, #{e}"
           end
+        else
+          raise RuntimeError, "Invalid key path"
         end
-      elsif File.file?(@key_path)
-        begin
-          data = read_key(@key_path)
-          @key_data << data if valid_key?(data)
-        rescue StandardError => e
-          @error_list << "#{@key_path} could not be read, #{e}"
+      end
+
+      if @private_key.present?
+        data = Net::SSH::KeyFactory.load_data_private_key(@private_key).to_s
+        if valid_key?(data)
+          @key_data << data
+        else
+          raise RuntimeError, "Invalid private key"
         end
-      else
-        raise RuntimeError, "No key path"
       end
+
+      unless @private_key.present? || @key_path.present?
+        raise RuntimeError, "No key path or key provided"
+      end
```

Comment:
Consider moving this guard to be the first tested condition.

Reply:
Agreed, always best to fail faster. Will make this change in the upcoming commit.